### ISSUE TITLE
Allow registration of custom Promise implementation in Browser Version

### ISSUE THREAD

**Issue by tnguyen14** [kevinbeaty/any-promise#11]
I am not sure if I am using this library wrong, but I'm getting this error when running a test on the library I'm writing:

```
Error: TypeError: 'undefined' is not a constructor (evaluating 'new Promise')
```

I am using this as suggested in the README inside my library code:

```js
var Promise = require('any-promise');
```

Do I have to register a promise implementation in the test code? Shouldn't it just fall back to native implementation? (I'm running on node v5.x)

**Comment by addaleax** [comment]
> Do I have to register a promise implementation in the test code? Shouldn't it just fall back to native implementation? (I'm running on node v5.x)

You’re right, it should work. Can you provide a stack trace/code or some other way to reproduce this?

**Comment by tnguyen14** [comment]
@addaleax I just realized that this might possibly be a problem when running with [zuul](https://github.com/defunctzombie/zuul), which is what the project I'm contributing to is using.

Here's a simple `test.js` file

```js
var test = require('tape');
var Promise = require('any-promise');

test('Promise should not be undefined', function (t) {
	t.plan(1);
	t.ok(Promise);
});
```

and then run it with `zuul --phantom -- test.js`.

**Comment by tnguyen14** [comment]
It could be that Promise is not available on the phantomjs browser. How should I load in a default promise implementation in the library?

**Comment by kevinbeaty** [comment]
I haven't used PhantomJS yet.  Will it use the `browserify` version?  

You can check by logging the `any-promise` implementation in your test:

```js
var impl = require('any-promise/implementation');
console.log(impl);
```

Does this log `window.Promise` or `global.Promise`?

Currently the browser version requires a pollyfill if a native Promise is not available.  We could change it so that `require('any-promise/register')('my-promise-lib')` would work if that were helpful.

**Comment by tnguyen14** [comment]
`impl` is returning as `"window.Promise"`

Is there any way to declare the polyfill at the library level, or must it be declared at the application level?

**Comment by kevinbeaty** [comment]
That makes sense.  Thanks for checking.

Currently the polyfill must be declared application-wide (ie, set `window.Promise`) for the browser version.

In the "normal" version, `require('any-promise/register')`  would register at the application level using a global variable, but not clobbering `global.Promise` per #8 .  

Although the concern with duplicate implementations may not be an issue in a browser. We could potentially support registering an implementation for the browser version and perhaps turning off global registration.  It would probably take the form of options in the register function containing an implementation and global registration flag.

```js
// Current (without registration)
var Promise = require('any-promise');  // loads global.Promise or window.Promise

// Current (with registration)
require('any-promise/register')('bluebird');  // call once at "entry point"
var Promise = require('any-promise'); // loads Bluebird or window.Promise (ignored in browser)

// Potential future release (optional options object for advanced use, backwards compatible)
require('any-promise/register')('mypromise', {implementation: require('my-promise'), global: false})
```

I'd have to think about this.  I haven't really thought about what should happen on the browser

What's your ideal behavior?

**Comment by tnguyen14** [comment]
@kevinbeaty just to clarify, you're asking for ideal behavior from the library perspective, or the application perspective? The documentation separates those 2 use cases, so I want to make sure.

**Comment by kevinbeaty** [comment]
I'm mainly just looking for context.  The link to the related issue helps provide me some.

I'd like to leave the behavior from the "library perspective" as-is.   I think the main benefit of using `any-promise` over just requiring a polyfill (or environment with a Promise implementation) is allowing the application author to override the default choice (and removing the need for the library author to make it for them).

The current behavior in the browser is to require a polyfill, so my suggestion above is to modify the browser version of `register` to work in the browser.   As far as I understand, we can't just do it with the current signature of `require('any-promise/register')('bluebird')` because using `require` with a variable (like `register.js`) does not play well with `browserify`.

So my suggestion above is to modify `register` to allow explicitly requiring the implementation with an optional options object.  This would allow registration in the browser version if the "application author" desires something other than `window.Promise` (it could also allow providing a decorated/custom Promise library instead of forcing a hidden `require` in `register.js`).

The `global` option seems to be a good idea too to allow separate JavaScript bundles without global registration.

It makes `register` a little more complex, but it only has to be called once, and only if the application author wants to change the default choice.   The library author still always just calls `var Promise = require('any-promise')`

Does something like that seem to address your needs, or is there something else you were thinking?

**Comment by tnguyen14** [comment]
@kevinbeaty apologies for the slow response. 

I think what you're describing makes sense. 

Would you like to create a branch to try this out? I could also start something, though I'm not 100% clear on what it is you have in mind.

**Comment by kevinbeaty** [comment]
@tnguyen14 A PR would definitely be appreciated as I've been swamped lately and not sure when I'll be able to get to this.

I think the first step (and maybe the only one you would need for your use case) is to modify the `register` function to accept an options object with the Promise implementation.

Something like this (untested):



In `register.js` (The Node.js version)

```js
module.exports = register
function register(implementation, opts){
  implementation = implementation || null
  opts = opts || {}

  // load any previous global registration
  if(registered === null){
    registered = global[REGISTRATION_KEY] || null
  }

  if(registered !== null
      && implementation !== null
      && registered.implementation !== implementation){
    // Throw error if attempting to redefine implementation
    throw new Error('any-promise already defined as "'+registered.implementation+
      '".  You can only register an implementation before the first '+
      ' call to require("any-promise") and an implementation cannot be changed')
  }

  if(registered === null){
    // load implementation
    if(implementation !== null){
      if(typeof opts.Promise !== 'undefined'){
         registered = {
           Promise: opts.Promise,
           implementation: implementation
         }
      } else {
         //... rest of implementation as is
```

In `regsiter-shim.js` (The Browser version), the implementation would be similar to `register.js`, but with no "auto-detect" or `require` with a lib (need to enforce opts.Promise in browser if want to use custom implementation).  Again, untested

```js
module.exports = register
function register(implementation, opts){
  implementation = implementation || null
  opts = opts || {}

  // load any previous global registration
  if(registered === null){
    registered = window[REGISTRATION_KEY] || null
  }

  if(registered !== null
      && implementation !== null
      && registered.implementation !== implementation){
    // Throw error if attempting to redefine implementation
    throw new Error('any-promise already defined as "'+registered.implementation+
      '".  You can only register an implementation before the first '+
      ' call to require("any-promise") and an implementation cannot be changed')
  }

  if(registered === null){
    // load implementation
    if(implementation !== null){
      if(typeof opts.Promise !== 'undefined'){
         registered = {
           Promise: opts.Promise,
           implementation: implementation
         }
      } else {
          throw new Error("Must specify Promise to load in browser, e.g. `require('any-promise/register')('bluebird', {Promise: require('bluebird').Promise}`")
     }
  } else {
     // fallback to window.Promise
     registered = {
       Promise: window.Promise,
       implementation = 'window.Promise'
     } 
  }
  window[REGISTRATION_KEY] = registration
}
  if(registered === null){
    throw new Error('Cannot find any-promise implementation nor'+
      ' window.Promise. You must install polyfill or call'+
      ' require("any-promise/register") with your preferred'+
      ' implementation, e.g. require("any-promise/register")("bluebird", {Promise: require('bluebird').Promise})'+
      ' on application load prior to any require("any-promise").')
  }

  return registered
}
```

**Comment by kevinbeaty** [comment]
The second thing I mentioned would be to allow bypassing global registration.  This could be potentially be useful for creating browser bundles without modifying the global scope.  I think this one is lower priority and probably less useful, but it might be nice (I'd be OK without this change if you are not interested).

This would also be a flag on the optional options object.

```js
function register(implementation, opts){
  implementation = implementation || null
  opts = opts || {}
  // global registration unless explicitly  {global: false} in options (default true)
  var registerGlobal = opts.global !== false;
...
```
Then everywhere you check for `global[REGISTRATION_KEY]` or `window[REGISTRATION_KEY]` you would guard with `if(registerGlobal){`.

**Comment by kevinbeaty** [comment]
Published in v1.2.0
